ci_tutorial/products_pipeline/full_pipeline.rst

```diff
@@ -0,0 +1,303 @@
+    $ conan graph build-order --requires=mapviewer/1.0 --lockfile=conan.lock 
+      --build=missing -s build_type=Debug --order-by=recipe --format=json > mapviewer_debug.json
+
+Likewise the ``build-order-merge`` command will be identical to the previous one.
+In this case, as this command doesn't really compute a dependency graph, a ``conan.lock`` argument is not necessary,
+dependencies are not being resolved:
+
+
+.. code-block:: bash
+
+    $ conan graph build-order-merge 
+      --file=game_release.json --file=game_debug.json 
+      --file=mapviewer_release.json --file=mapviewer_debug.json 
+      --reduce --format=json > build_order.json
+
+
+    
+So far, this process has been almost identical to the previous section one, just with the difference of capturing and using a lockfile.
+Now, we will explain the "core" of the ``products`` pipeline: iterating the build-order and distributing the build, gathering the 
+resulting built packages.
+
+This would be some Python code that performs the iteration sequentially (a real CI system would distribute the builds to different agents in parallel):
+
+
+.. code-block:: python
+
+  build_order = open("build_order.json", "r").read()
+  build_order = json.loads(build_order)
+  to_build = build_order["order"]
+
+  pkg_lists = []  # to aggregate the uploaded package-lists
+  for level in to_build:
+      for recipe in level:  # This could be executed in parallel
+          ref = recipe["ref"]
+          # For every ref, multiple binary packages are being built. 
+          # This can be done in parallel too. Often it is for different platforms
+          # they will need to be distributed to different build agents
+          for packages_level in recipe["packages"]:
+              # This could be executed in parallel too
+              for package in packages_level:
+                  build_args = package["build_args"]
+                  filenames = package["filenames"]
+                  build_type = "-s build_type=Debug" if any("debug" in f for f in filenames) else ""
+                  run(f"conan install {build_args} {build_type} --lockfile=conan.lock --format=json", file_stdout="graph.json")
+                  run("conan list --graph=graph.json --format=json", file_stdout="built.json")
+                  filename = f"uploaded{len(pkg_lists)}.json"
+                  run(f"conan upload -l=built.json -r=products -c --format=json", file_stdout=filename)
+                  pkg_lists.append(filename)
+
+
+.. note::
+
+  - This code is specific for the ``--order-by=recipe`` build-order, if chosing the ``--order-by=configuration``, the json
+    is different and it would require a different iteration.
+
+
+These are the tasks that the above Python code is doing:
+
+- For every ``package`` in the build-order, a ``conan install --require=<pkg> --build=<pkg>`` is issued, and the result of this command is stored in a ``graph.json`` file
```

Comment:
Is there a special reason, why `conan install`is used here, instead of `conan create`?
I'm also asking because `conan install` does not seem to execute `test`...

Reply:
Yes, building from Conan recipes in the server is way simpler. Otherwise it is necessary to obtain the exact ``scm`` coordinates (recipes must implement the scm-capture strategy), do the clone+checkout then the ``conan create``. 

It has the advantage that it runs the ``test_package``, but this ``test_package`` is focused on checking the package correctness (the headers have been packaged, the libs too, etc), it is not a functional test. As in a CI pipeline packages are being re-built, but their code or recipes is not changed at all, it is almost impossible that the ``test_package`` would catch any significant package issue, as they are extremely unlikely. And the regular testing that happens in the ``build()`` when activated, and the testing performed by the consumers of packages that have been re-built is even a better check of the package correctness, as they are the real consumers.

Reply:
We will most likely add examples in a later section of the docs with the full ``scm`` + clone + checkout + ``conan create`` flow. Specially because it will most likely be highly simplified by new ``conan workspace`` commands.

But at this point, it is better to focus on the basic concepts of the CI process, the simpler the examples and commands the better.